### PR TITLE
mitigate thundering herd problems

### PR DESCRIPTION
**Pull request by jmhodges**
This allows a single ctile process to turn multiple simultaneous
requests for the same tile into a single request of each kind needed to
the backend. This request collapsing can be a pretty handy technique for
a complex caching system like ctile. We use the singleflight library to
do so.

This doesn't collapse requests across different ctile processes. That
would require more distributed systems work (using groupcache or
whatever). The assumption here is that there will be relatively few
ctile processes running in the cluster, so the thundering herd problem
being mitigated in a single process is useful and a few hosts making
singular simultaneous requests is currently an okay cost.

Future work to do request collapsing across ctile instances will be
easier to do with this singleflight joint in place.

We perform this request collapsing across all of the remote operations:
the initial fetch attempt from S3, the fetching from the backing CT log,
and writing to S3. Including all 3 operations at once avoids certain
race conditions. For instance, if we'd not included the initial S3
fetch, we'd hit the classic bug where two threads see that a cache entry
is empty, but one finishes filling the cache before the second has even
started the work to fill it and so the system performs the cache filling
a second time.

There is one minor caveat that every request coming into the same
singleflight.Do will have the same deadline as the request that first
requested the singleflight.Do. If something had happened to eat a lot of
time on the first request (say, garabage collection), but not the later
requests, a timeout could be returned for all of them. This is an
unlikely scenario but a possibility.

Future work could mitigate that by checking if the error returned from
the singleflight.Do was context.DeadlineExceeded while the local
goroutine's `ctx.Error` is nil and retrying if so. That felt like change
that would muddle this PR, but it feels reasonable to do later.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -229,11 +230,13 @@ func (tch *tileCachingHandler) getFromS3(ctx context.Context, t tile) (*entries,
...
 type tileCachingHandler struct {
 	logURL   string // The string form of the HTTP host and path prefix to add incoming request paths to in order to fetch tiles from the backing CT log. Must not be empty.
-	tileSize int    // The CT tile size used here and in the given backend. Must not be zero.
+	tileSize int    // The CT tile size used here and in the backing CT log. Must be the same as the backing CT log's value and must not be zero.
 
 	s3Service *s3.Client // The S3 service to use for caching tiles. Must not be nil.
 	s3Prefix  string     // The prefix to add to the path when caching tiles in S3. Must not be empty.
 	s3Bucket  string     // The S3 bucket to use for caching tiles. Must not be empty.
+
+	cacheGroup *singleflight.Group // The singleflight.Group to use for dedupulicating simultaneous requests (a.k.a. "request collapsing") for tiles. Must not be nil.
```
Comment (jcjones) [inline main.go]:
```suggestion
	cacheGroup *singleflight.Group // The singleflight.Group to use for deduplicating simultaneous requests (a.k.a. "request collapsing") for tiles. Must not be nil.
```

**Review by jcjones — COMMENTED** [review]
drive by spelling attaaaacccckkkkkk

File: main.go [hunk 2]
```diff
@@ -310,7 +313,32 @@ const (
...
+// request. It should be preferred over getAndCacheTileUncollapsed.
 func (tch *tileCachingHandler) getAndCacheTile(ctx context.Context, tile tile) (*entries, tileSource, error) {
+	dedupKey := fmt.Sprintf("logURL-%s-tile-%d-%d", tile.logURL, tile.start, tile.end)
+
+	type entriesAndSource struct {
+		entries *entries
+		source  tileSource
+	}
+
+	innerContents, err, _ := singleflightDo(tch.cacheGroup, dedupKey, func() (entriesAndSource, error) {
```
Comment (jcjones) [inline main.go:328]: It'd be pretty cool as a follow-on to collect arg3 here (`shared bool`) for metrics.

Comment (jcjones) [reply main.go:328]: https://github.com/letsencrypt/ctile/issues/10

Comment (jmhodges) [reply main.go:328]: Yep, could be handy!

**Review by jsha — APPROVED** [review]
Nice!

**Pull request merged**